### DISCUSSION TITLE
The way to build latest Android APK in MacOS Android Studio

### DISCUSSION THREAD

**Discussion by linonetwo** [CleverRaven/Cataclysm-DDA#46249] [Show and tell]
The Android build always fails, so we should build it by our self.

## Open the correct folder

Open the `android` folder in the Android Studio

<img width="752" alt="截屏2020-12-23 下午1 44 26" src="https://user-images.githubusercontent.com/3746270/102963821-139b8d80-4525-11eb-8113-02ff92cd07c7.png">

## install things

<img width="806" alt="A5FB3405F748743B572EB8D074C311D4" src="https://user-images.githubusercontent.com/3746270/102963943-5e1d0a00-4525-11eb-9cf3-0a8c853a861e.png">

Click the elephant icon

## set versionCode

First, you can change `android/app/build.gradle`, so you can make sure you won't encounter "can not downgrade your app" when installing (see https://developer.android.com/studio/publish/versioning for details)

```diff
-       versionCode Integer.valueOf(System.env.UPSTREAM_BUILD_NUMBER ?: 1)
+       versionCode Integer.valueOf(System.env.UPSTREAM_BUILD_NUMBER ?: 11281)
```

The number here should change to the latest on in https://cataclysmdda.org/experimental/ , for now, it is 11281

<img width="833" alt="截屏2020-12-22 下午6 25 20" src="https://user-images.githubusercontent.com/3746270/102878356-0e860200-4483-11eb-9f1f-85986a827e51.png">

## Set versionName

In `android/gradle.properties`, set a friendly number for your app's version. So you can see the version number when you install your .apk

```diff
override_version=
override_version=11281
```

## Build

<img width="204" alt="350F08F11B01917630FDAF726F81FE84" src="https://user-images.githubusercontent.com/3746270/102963991-7f7df600-4525-11eb-91a1-0f2c4a706845.png">

<img width="507" alt="截屏2020-12-23 下午1 48 38" src="https://user-images.githubusercontent.com/3746270/102964019-9290c600-4525-11eb-99c7-7abf19cdcb29.png">

This can be very slow, on my high-end MacOS, it still takes 15min, on my friend's small windows laptop, it takes a whole night.

<img width="445" alt="CE6618DFEE6B290F64B243FF21EDE921" src="https://user-images.githubusercontent.com/3746270/102964003-87d63100-4525-11eb-86bb-34450303dce0.png">

## Change signing

It will say Signing is different from existed app upon installing, sadly we can't change our signing to the same as the official one download from https://cataclysmdda.org/experimental/ or http://dev.narc.ro/cataclysm/jenkins-latest/Android/Tiles/

So we can't just overwrite install on an existing CDDA app. We have to

1. backup data in Android/data/com.CleverRaven.Cataclysm-DDA.experimental (Backup android data in macOS will be very slow, so you may want to use https://github.com/syncthing/syncthing to back it up to the computer, you can do this by temporarily enable single direction sync)
2. uninstall the old experimental app
3. install new app build by your self
4. restore data (maybe by single direction sync, or manually copy and paste)

> You can also change applicationId in Gradle file before doing your local build, so CleverRaven version and your version could be installed side-by-side and won't interfere whatsoever.

We can sync save between two apps using SyncThing, as it can sync data between several folders. In this way, we can use our won built app if the official one is not released for a long time, and use the official one if it gets released.

**Comment by actual-nh** [comment]
Always fails?

**Reply by linonetwo** [reply]
Okay, this is an exaggeration, but look at http://dev.narc.ro/cataclysm/jenkins-latest/Android/Tiles/ it fails 9 out of 10, nearly only release 1 time every 10 days.

An expert said it is because out of disk space https://github.com/CleverRaven/Cataclysm-DDA/issues/45770#issuecomment-744227665
